nx/lib/nx/defn/kernel.ex

```diff
@@ -1013,4 +1013,129 @@ defmodule Nx.Defn.Kernel do
   defmacro @expr do
     quote do: Kernel.@(unquote(expr))
   end
+
+  @doc """
+  Asserts the `tensor` has a certain `shape`.
+
+  If it succeeds, it returns the given tensor. Raises
+  an error otherwise.
+
+  ## Examples
+
+  To assert the tensor is a scalar, you can pass the empty tuple `shape`:
+
+      iex> assert_shape Nx.tensor(13), {}
+      #Nx.Tensor<
+        s64
+        13
+      >
+
+  If the shapes do not match, an error is raised:
+
+      iex> assert_shape Nx.tensor([1, 2, 3]), {}
+      ** (ArgumentError) expected tensor to be a scalar, got tensor with shape {3}
+
+      iex> assert_shape Nx.tensor([1, 2, 3]), {4}
+      ** (ArgumentError) expected tensor to have shape {4}, got tensor with shape {3}
+
+  If you want to assert on the rank or shape patterns, use
+  `assert_shape_pattern/2` instead.
+  """
+  def assert_shape(tensor, shape) when is_tuple(shape) do
+    case Nx.shape(tensor) do
+      ^shape ->
+        tensor
+
+      other ->
+        raise ArgumentError,
+              "expected tensor to #{shape_to_string(shape)}, got tensor with shape #{inspect(other)}"
+    end
+  end
+
+  defp shape_to_string({}), do: "be a scalar"
+  defp shape_to_string(shape), do: "have shape " <> inspect(shape)
+
+  @doc """
+  Asserts the `tensor` has a certain `shape` pattern.
+
+  If it succeeds, it returns the given tensor. Raises
+  an error otherwise.
+
+  ## Examples
+
+  Opposite to `assert_shape/2`, where the given shape is a value,
+  `assert_shape_pattern` allows the shape to be any Elixir pattern.
+  We can use this to match on ranks:
+
+      iex> assert_shape_pattern Nx.tensor([[1, 2], [3, 4]]), {_, _}
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [1, 2],
+          [3, 4]
+        ]
+      >
+
+      iex> assert_shape_pattern Nx.tensor([1, 2, 3]), {_, _}
+      ** (ArgumentError) expected tensor to match shape {_, _}, got tensor with shape {3}
+
+  Or even use variables to assert on properties such as square matrices:
+
+      iex> assert_shape_pattern Nx.tensor([[1, 2], [3, 4]]), {x, x}
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [1, 2],
+          [3, 4]
+        ]
+      >
+
+      iex> assert_shape_pattern Nx.tensor([1, 2, 3]), {x, x}
+      ** (ArgumentError) expected tensor to match shape {x, x}, got tensor with shape {3}
+
+  You can also use guards to specify tall matrices and so forth:
+
+      iex> assert_shape_pattern Nx.tensor([[1], [2]]), {x, y} when x > y
+      #Nx.Tensor<
+        s64[2][1]
+        [
+          [1],
+          [2]
+        ]
+      >
+
+      iex> assert_shape_pattern Nx.tensor([1, 2]), {x, y} when x > y
+      ** (ArgumentError) expected tensor to match shape {x, y} when x > y, got tensor with shape {2}
+
+  """
+  defmacro assert_shape_pattern(tensor, shape) do
+    shape_pattern_string = shape_pattern_to_string(shape)
+
+    quote do
+      tensor = unquote(tensor)
```

Comment:
Any reason not to use `:bind_quoted`?

Reply:
bind_quoted disables unquoting and we need to unquote the shape inside the pattern position. So I would need to pass both bind_quoted and enable unquoting and that would be more verbose. :)

Reply:
Ah right, thanks!